### PR TITLE
autoscaling: add querier predictive scaling

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
This PR upstreams a change that we have been trialing at GL.

### Problem

Some workloads have a very uneven distribution of queries. Some of the causes of this can be external alerting systems and reporting tools which run batch jobs once a day.

This results in an influx of queries at fairly regular intervals (~24h). Autoscaling can only respond so fast to incoming queries. In practice, it takes anywhere between 2 and 7 minutes to respond to queries and trigger a scaling event. As a result Mimir's query queues fill up and Mimir start returning HTTP 429 to new queries until queriers have had the time to scale out and accommodate the added load.

### Solution

Scale preemptively. Automated workloads are also highly predictive. This PR uses the value of the inflight queries at the same time last week (6d23h30m instead of 7d) to prepare for periodic queries.

### Results

We've seen HTTP 429 responses decrease significantly on a weekly basis: between 66% and 95% 
Screenshot 2024-04-02 at 15 13 41

### Side effects

There are a lot of false-positive scaling events. As a result of this we've seen querier cost increase between 20% and 50%. Given that querier cost is usually 1-2% of the cell's TCO, this is somewhat acceptable for a better QoS.

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by dimitarvdimitrov** [comment]
> Is it worth making the lookback (currently 7d) and range (currently 30m) configurable? I could imagine different Mimir clusters could have different query patterns.

Added in https://github.com/grafana/mimir/pull/7775/commits/a4ba23bb435b6d98abb31b95486262093a292b67

File: operations/mimir/autoscaling.libsonnet [hunk 1]
```diff
@@ -6,6 +6,9 @@
     autoscaling_querier_min_replicas: error 'you must set autoscaling_querier_min_replicas in the _config',
     autoscaling_querier_max_replicas: error 'you must set autoscaling_querier_max_replicas in the _config',
     autoscaling_querier_target_utilization: 0.75,  // Target to utilize 75% querier workers on peak traffic, so we have 25% room for higher peaks.
+    autoscaling_querier_predictive_scaling_enabled: false,  // Use inflight queries from 7d ago to predict the number of queriers needed.
+    autoscaling_querier_predictive_scaling_period: '6d23h30m',  // The period to consider when considering scheduler metrics for predictive scaling. This is usually slightly lower than the period of the repeating query events to give scaling up head time.
+    autoscaling_querier_predictive_scaling_lookback: '30m',  // The time range to consider when considering scheduler metrics for predictive scaling. For example: if lookback is 30m and period is 6d23h30m, the querier will scale based on the maximum infligth queries between 6d23h00m and 6d23h30m ago.
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply operations/mimir/autoscaling.libsonnet:11]: (same comment and typo applies to Helm chart too)

File: operations/mimir/autoscaling.libsonnet [hunk 2]
```diff
@@ -6,6 +6,9 @@
     autoscaling_querier_min_replicas: error 'you must set autoscaling_querier_min_replicas in the _config',
     autoscaling_querier_max_replicas: error 'you must set autoscaling_querier_max_replicas in the _config',
     autoscaling_querier_target_utilization: 0.75,  // Target to utilize 75% querier workers on peak traffic, so we have 25% room for higher peaks.
+    autoscaling_querier_predictive_scaling_enabled: false,  // Use inflight queries from 7d ago to predict the number of queriers needed.
+    autoscaling_querier_predictive_scaling_period: '6d23h30m',  // The period to consider when considering scheduler metrics for predictive scaling. This is usually slightly lower than the period of the repeating query events to give scaling up head time.
```
Comment (charleskorn) [inline operations/mimir/autoscaling.libsonnet:10]: Not sure about the use of "head time" here - perhaps "lead time" would be clearer?

Comment (dimitarvdimitrov) [reply operations/mimir/autoscaling.libsonnet:10]:
```suggestion
    autoscaling_querier_predictive_scaling_period: '6d23h30m',  // The period to consider when considering scheduler metrics for predictive scaling. This is usually slightly lower than the period of the repeating query events to give scaling up lead time.
```

File: operations/helm/charts/mimir-distributed/values.yaml [hunk 3]
```diff
@@ -1286,6 +1286,17 @@ querier:
...
+    # -- predictiveScalingEnabled scales up the querier based on the inflight requests a week ago (6d23h30m).
+    # This helps with scaling up for predictable traffic patterns and minimizing HTTP 429 responses due to filled query queues.
+    # Due to false positive items it can increase the querier TCO.
+    predictiveScalingEnabled: false
+    # -- The period to consider when considering scheduler metrics for predictive scaling.
+    # This is usually slightly lower than the period of the repeating query events to give scaling up head time.
+    predictiveScalingPeriod: 6d23h30m
+    # -- The time range to consider when considering scheduler metrics for predictive scaling.
+    # For example: if lookback is 30m and period is 6d23h30m,
+    # the querier will scale based on the maximum infligth queries between 6d23h00m and 6d23h30m ago.
```
Comment (dimitarvdimitrov) [inline operations/helm/charts/mimir-distributed/values.yaml:1298]:
```suggestion
    # the querier will scale based on the maximum inflight queries between 6d23h30m and 7d ago.
```

**Comment by dimitarvdimitrov** [comment]
thanks for the review and sorry for the numerous typos 🙈 can you take another pass @charleskorn?

File: operations/helm/charts/mimir-distributed/templates/querier/querier-so.yaml [hunk 4]
```diff
@@ -44,4 +44,15 @@ spec:
       {{- end }}
     name: cortex_querier_hpa_default_requests_duration
     type: prometheus
-{{- end }}
\ No newline at end of file
+  {{- if .Values.querier.kedaAutoscaling.predictiveScalingEnabled }}
+  - metadata:
+      query: sum(max_over_time(cortex_query_scheduler_inflight_requests{container="query-scheduler",namespace="{{ .Release.Namespace }}",quantile="0.5"}[30m] offset 6d23h30m))
```
Comment (charleskorn) [inline operations/helm/charts/mimir-distributed/templates/querier/querier-so.yaml:49]: Shouldn't this use the values of `predictiveScalingPeriod` and `predictiveScalingLookback`?

File: operations/helm/charts/mimir-distributed/values.yaml [hunk 5]
```diff
@@ -1286,6 +1286,17 @@ querier:
     minReplicaCount: 1
     maxReplicaCount: 10
     querySchedulerInflightRequestsThreshold: 12
+    # -- predictiveScalingEnabled scales up the querier based on the inflight requests a week ago (6d23h30m).
+    # This helps with scaling up for predictable traffic patterns and minimizing HTTP 429 responses due to filled query queues.
+    # Due to false positive items it can increase the querier TCO.
+    predictiveScalingEnabled: false
+    # -- The period to consider when considering scheduler metrics for predictive scaling.
+    # This is usually slightly lower than the period of the repeating query events to give scaling up head time.
```
Comment (charleskorn) [inline operations/helm/charts/mimir-distributed/values.yaml:1294]:
```suggestion
    # This is usually slightly lower than the period of the repeating query events to give scaling up lead time.
```

File: operations/helm/charts/mimir-distributed/values.yaml [hunk 6]
```diff
@@ -1286,6 +1286,17 @@ querier:
     minReplicaCount: 1
     maxReplicaCount: 10
     querySchedulerInflightRequestsThreshold: 12
+    # -- predictiveScalingEnabled scales up the querier based on the inflight requests a week ago (6d23h30m).
```
Comment (charleskorn) [inline operations/helm/charts/mimir-distributed/values.yaml:1289]: I think this is more accurate now that the time range is configurable:

```suggestion
    # -- predictiveScalingEnabled scales up the querier based on the inflight requests in the past.
```

File: operations/mimir/autoscaling.libsonnet [hunk 7]
```diff
@@ -6,6 +6,9 @@
     autoscaling_querier_min_replicas: error 'you must set autoscaling_querier_min_replicas in the _config',
     autoscaling_querier_max_replicas: error 'you must set autoscaling_querier_max_replicas in the _config',
     autoscaling_querier_target_utilization: 0.75,  // Target to utilize 75% querier workers on peak traffic, so we have 25% room for higher peaks.
+    autoscaling_querier_predictive_scaling_enabled: false,  // Use inflight queries from 7d ago to predict the number of queriers needed.
```
Comment (charleskorn) [inline operations/mimir/autoscaling.libsonnet:9]: Same as above, I think this is more accurate now that the time range is configurable:

```suggestion
    autoscaling_querier_predictive_scaling_enabled: false,  // Use inflight queries from the past to predict the number of queriers needed.
```

File: operations/helm/charts/mimir-distributed/templates/querier/querier-so.yaml (hunk 4 above)
Comment (dimitarvdimitrov) [reply operations/helm/charts/mimir-distributed/templates/querier/querier-so.yaml]: it is now

**Review by charleskorn — APPROVED** [review]

**Review by pracucci — APPROVED** [review]
Nice work! Have you considered enabling it in `operations/mimir-tests/test-autoscaling.jsonnet`?

**Pull request merged**